// @decky/ui@3.26.2/dist/webpack.js
import Logger from './logger';
const logger = new Logger('Webpack');
export let modules = [];
function initModuleCache() {
    const startTime = performance.now();
    logger.group('Webpack Module Init');
    const id = Math.random();
    let webpackRequire;
    window.webpackChunksteamui.push([
        [id],
        {},
        (r) => {
            webpackRequire = r;
        },
    ]);
    logger.log('Initializing all modules. Errors here likely do not matter, as they are usually just failing module side effects.');
    for (let i of Object.keys(webpackRequire.m)) {
        try {
            const module = webpackRequire(i);
            if (module) {
                modules.push(module);
            }
        }
        catch (e) {
            logger.debug('Ignoring require error for module', i, e);
        }
    }
    logger.groupEnd(`Modules initialized in ${performance.now() - startTime}ms...`);
}
initModuleCache();
export const findModule = (filter) => {
    for (const m of modules) {
        if (m.default && filter(m.default))
            return m.default;
        if (filter(m))
            return m;
    }
};
export const findModuleDetailsByExport = (filter, minExports) => {
    for (const m of modules) {
        if (!m)
            continue;
        for (const mod of [m.default, m]) {
            if (typeof mod !== 'object')
                continue;
            if (minExports && Object.keys(mod).length < minExports)
                continue;
            for (let exportName in mod) {
                if (mod?.[exportName]) {
                    const filterRes = filter(mod[exportName], exportName);
                    if (filterRes) {
                        return [mod, mod[exportName], exportName];
                    }
                    else {
                        continue;
                    }
                }
            }
        }
    }
    return [undefined, undefined, undefined];
};
export const findModuleByExport = (filter, minExports) => {
    return findModuleDetailsByExport(filter, minExports)?.[0];
};
export const findModuleExport = (filter, minExports) => {
    return findModuleDetailsByExport(filter, minExports)?.[1];
};
export const findModuleChild = (filter) => {
    for (const m of modules) {
        for (const mod of [m.default, m]) {
            const filterRes = filter(mod);
            if (filterRes) {
                return filterRes;
            }
            else {
                continue;
            }
        }
    }
};
export const findAllModules = (filter) => {
    const out = [];
    for (const m of modules) {
        if (m.default && filter(m.default))
            out.push(m.default);
        if (filter(m))
            out.push(m);
    }
    return out;
};
export const CommonUIModule = modules.find((m) => {
    if (typeof m !== 'object')
        return false;
    for (let prop in m) {
        if (m[prop]?.contextType?._currentValue && Object.keys(m).length > 60)
            return true;
    }
    return false;
});
export const IconsModule = findModuleByExport((e) => e?.toString && /Spinner\)}\)?,.\.createElement\(\"path\",{d:\"M18 /.test(e.toString()));
export const ReactRouter = findModuleByExport((e) => e.computeRootMatch);
